### ISSUE TITLE
Newgrounds support

### ISSUE THREAD

**Issue by Simon-Tesla** [Simon-Tesla/RaccoonyWebEx#125]
Given some artists use Newgrounds as a gallery, it'd be good to get the plugin to recognize Newgrounds pages.

**Comment by Eupeptic1** [comment]
I have started working on this.  So far, I've been able to download three images by two artists.  :D  

In general, it's the old familiar story: there are several submission types, and the markup for each type varies a little, meaning the plugin has to look in different places to find everything for each submission type.

Newgrounds supports at least art (static images), audio, movies, and games.

Art: The plugin finds the main image, the thumbnail, and all the metadata.  Open-gallery-in-tabs is TODO.

Audio: Checked one.  I *think* there's a link to an mp3 available, but it may be buried inside a `<script>`, which I don't know if we can "see into".

Movies: Checked one.  It doesn't seem to load the video player by default; once you "click here to play", there's a URL to an .mp4 available in the page source.

Games: Newgrounds used to be the home of Flash stuff, but it looks like now people are using HTML5 "game engines".  Not sure if it's possible to download the entire game as one file, or if the file on Newgrounds also depends on things hosted elsewhere.

**Comment by Eupeptic1** [comment]
I worked on it a little more tonight.  This is a status update only; nothing earth-shattering.

Open-in-tabs works from the top-level "favorites" page ( user.newgrounds.com/favorites ) and from the type-specific favorites page ( user.newgrounds.com/favorites/(movies|games|audio|art) ), at least for the initial list of items.

The plugin can now find the metadata for movies as well.  It can't download the movie yet.

Discoveries:

The type-specific favorites have infinite scroll.  :(  It helps to look at a user with lots of favorites to see this, such as zellas.newgrounds.com .  Right now, I only see the set that is initially loaded (about 40 artworks or 50 songs); a mutation selector my be in my future.

For movies, you can't derive the direct URL to the .mp4 from the "click to play" screen you get at first.  You have to "click to play" to get the URL.  (This is probably on purpose, from Newgrounds' point of view.)  This may also need a mutation selector, so we can run again after the link is available.

**Comment by Eupeptic1** [comment]
I worked on it a little more tonight.  It's not done, but it's far enough along that I uploaded it to my repo, on the newgrounds-initial-support branch.  Code for the new plugin is here: https://github.com/Eupeptic1/RaccoonyWebEx/blob/newgrounds-initial-support/app/plugins/newgrounds.ts

### Questions for Simon

1.  Performance.  I added a mutation selector on body.  It works OK for handling infinite scroll, but it seems to cause a lot of calls of the plugin when watching a video.  If you go and sit on the "click to play" page for a video, the plugin runs once or twice and then stops.  Once you click "play", though, the plugin gets re-run repeatedly; I don't think it's every video frame but it's around 1 Hz or a little faster.  It doesn't seem to make the movie skip or stutter, but it seems like it's running frequently.  Is this potentially bad?

Also, If I just go and "park" on an artwork page, the plugin re-runs once every 10 seconds.  I'm not sure why; maybe Newgrounds is trying to update an ad, or updating the comments/favorites/rating, or something like that.  I am not as worried about this one but I figured I'd note it.

2.  Getting the .mp3 link for audio.  Right now I find a particular inline script that has the URL in it, and parse the URL out of the script's `innerHTML`.  It works, but I have a feeling that might make your tail curl up.  :D  I _think_ I can build the URL from the submission ID, which I can find in the conventional way... would that be better?

3.  Do we want to support downloading games at all?  I'm not sure if it's possible but I haven't looked hard at it yet.

The rest of this is a status update and is not required reading.

#### Status update:

I added a mutation selector.  When you scroll to the bottom of a long favorites list, and Newgrounds loads another batch of favorites, the open-in-tabs code now reparses the page and finds the new, longer list of favorites.

I worked on the movie support.  Once the user is on the page for an individual movie, they have to click once on the "click to play" screen.  This loads Newgrounds' video player, which puts a link to the mp4 in the source.  The plugin will notice it and offer the raccoon-head menu shortly after the video starts playing.  It can correctly download the .mp4 and the metadata file.

I also worked on the audio support.  In contrast to movies, Newgrounds loads the embedded audio player right away, when you are on the page for an individual song.  There is a link to an .mp3 available, but it's inside an inline `<script>` tag.  I, uh, find the script and parse the link out of the script text.  <_<  (The alternative would be to generate the .mp3 link from the submission ID, which seems possible to do - Newgrounds doesn't seem to add any extra hashes/random numbers to the URL to prevent that.)

#### TODO list

The markup on some submission types might vary if there are multiple authors listed - there might be a `<ul>` that's not in there otherwise.  I need to find submissions of all types with one and with multiple authors to see how that works.

There is always a description in the OpenGraph meta tags for all submission types.  I have a small sample size, but for art, that description matches what the site shows below the picture.  For movies and audio, there is an "author comment" below the media player, which sometimes contains a longer or more informative comment than what's in the meta tag.  Right now, I always use what's in the meta tag, but it might be more useful to use the "author comment" for some media types.

The plugin is currently of the style where it tries `getMedia()` and `getPageLinkList()` on all pages, rather than only trying to run `getMedia()` on certain pages.  I may want to change that.

Make a list of test links that cover some of the combinations and permutations, and download the submissions in those links, as a base for future semi-automated testing.

**Comment by Simon-Tesla** [comment]
I just realized I should've looked at this issue before starting work, but I had an itch and started working on my own version without remembering you had been working on something. If you want to publish your changes we might be able to combine them.

That said, I did do some core code updates to make Raccoony a bit more friendly to Newgrounds, specifically to make the interaction a bit cleaner for movie support, as it will recognize that the page is a movie page from the start and show a message when you try to download without hitting play, and doesn't require a mutation selector at all. I didn't add audio support though, as I basically worked on this on a whim late last night and it looked like more trouble than I was willing to put up with.

**Comment by Simon-Tesla** [comment]
To answer questions
1. What might make sense here is to add a debounce for the mutation handler such that it only reruns the detection logic once per minute or something like that. It's not a lot of code to run but it's enough that I'd prefer to tamp it down if we do need it. As I mentioned above I made alterations to core Raccoony that should mean a mutation selector is unnecessary for Newgrounds.
2. That sounds fine to me for audio support. It's not awesome, but I don't see a better way of doing it either. It *might* be possible to [yank it from a variable](https://stackoverflow.com/questions/3955803/chrome-extension-get-page-variables-in-content-script) in the script but that might be more hassle than it's worth.
3. I'd consider games out of scope. If they're flash, then they're not easy to run, and if they're HTML5 they need to be run in a browser anyway. There's no real meaningful way to download an HTML5 game.

**Comment by Eupeptic1** [comment]
> If you want to publish your changes we might be able to combine them.

All of my Newgrounds stuff is on the newgrounds-initial-support branch in my repo: https://github.com/Eupeptic1/RaccoonyWebEx/tree/newgrounds-initial-support

I *think* I branched from vanilla 1.2.5. I haven't touched any of it since mid-May, 2022.

If you need me to submit a PR or similar to make it easier to combine my code with your new code, please let me know what I should do.

**Request:** If you find some submissions at Newgrounds that you think might make good test cases (multiple authors, funny characters in title/author/description, claims to be an `.mp3` but turns out to be a `.wav`, and similar), bookmark them. I don't have any test cases for Newgrounds yet and I would like to have some.

Thanks!

**Comment by Simon-Tesla** [comment]
Github's made a lot of strides making comparing branches in other user repos easier, so I was able to incorporate the code relatively easily.

As far as test cases I've kinda just been browsing around. The one I'm using for art is Very NSFW, so probably not great for use as a test case.

**Comment by Simon-Tesla** [comment]
I've just updated the PR to add audio support, support for the user feed, and for the favorites gallery. If you want to take a glance at it, feel free. I'll probably go ahead and check it in tomorrow.

**Comment by Eupeptic1** [comment]
I looked at the code but I haven't run it yet.  I don't want to try it now, but I may be able to run it later on this afternoon or evening.

On the test cases: given the nature of the sites we target, I think it'll be difficult to find a lot of submissions that both 1) have some structure that has known to be tricky to parse, or been a bug for us in the past and 2) are SFW. In a perfect world, we could check the test case list in along with the code, but in this vale of tears, I send it to you via email. :D

**Comment by Eupeptic1** [comment]
I grabbed the code from your `newgroundsPlugin` branch, built it, installed it, and started plinking around. 

#### tl;dr:

The real basics work: open-in-tabs, downloading most submissions, downloading metadata file if enabled.

If you're sitting on the static intro page of a movie, and click "Download", the "must press play" alert comes up correctly.

I saw a few probable bugs:

* the art support can't see WebP files (the page markup is different)
* "download opened submissions" didn't do anything when I did an open-all-in-tabs on a bunch of mp3 files
* "download opened submissions" missed some of the submissions for one artist I tried it on
* open-in-tabs doesn't respect Newgrounds' blacklist/unsuitable flags

All of this was in the logged-out condition, as I don't have a Newgrounds account yet.  I will make one soon, to support further testing.

#### The long version

#### Bugs:

1.  The plugin currently doesn't find WebP images.  Example, sfw: https://www.newgrounds.com/art/view/vinnx4/sonia-animated

Looking at it a little bit, the plugin is looking for an element with an id of `#portal_item_view` to find the image.  On a static image, the last part of the path to the image is `div.pod_body div.image a#portal_item_view`, so that works.  For a WebP, the last part of the path to the image is `div.pod_body div.image img`, so the plugin misses it.

I would *guess* the `portal_item_view` is related to the click-to-embiggen feature Newgrounds uses for static images; it doesn't offer that for WebP images.

2.  "Auto download opened submissions" doesn't seem to work at all, when opening a number of audio submissions.  Example, sfw: https://drakensson.newgrounds.com/audio/

I turned on auto-download, visited that page, didn't scroll down, and did an open-all-in-tabs.  Raccoony opened a new tab for each song, but nothing downloaded.

I also tried manually right-clicking on one of the songs in the gallery and selecting Firefox's "open in new tab" option; a new tab for that song opened, but Raccoony didn't download it.

Normally, on an audio page, there is a rectangle with the waveform of the song near the top.  When the open-all-in-tabs was running, I looked at one of the new tabs - when it first loaded, there was just a black rectangle with a small logo/initials in the upper left corner in that area; after a second or so, the waveform loaded in.  Maybe Raccoony can't find the song until that happens.

3.  "Auto download opened submissions" doesn't download some images, sometimes.

I was on this artist's page (sfw) https://andyl4nd.newgrounds.com/art and did an open-all-in-tabs.  Raccoony opened 23 tabs, but only downloaded the artwork for 13 of them.  The 13 that did download were a mix of PNG and GIF images.

The 10 that didn't download are also a mix of PNG and GIF images, according to the infoboxes on Newgrounds.  I checked a few of the GIFs out of the 10 that didn't download and they really do seem to be GIF files on Newgrounds' server - not "silently" converted to WebP or something.

I was doing something in another tab, unrelated to Raccoony, that crashed Firefox.  I restarted Firefox, reloaded Raccoony in `about:debugging`, and tried it again.  Same result: Raccoony only downloaded the same 13 files, and skipped the same 10 files.

4.  "Open in tabs" doesn't observe the blacklist.

While logged out of Newgrounds, I went to a user's page who had faved 2 art submissions.  Both submissions were rated "A"dult, so Newgrounds displayed a black thumbnail with a crossed-out eye, instead of the normal thumbnail, for both of them.  I then did an "open all in tabs", and Raccoony did so; Newgrounds displayed the artwork in each tab, without any kind of "restricted" or "please log in" message.

That same user had also faved 2 A-rated movies; I tried open-all-in-tabs there and got the same result. Raccoony opened a new tab for each movie.

This may require a similar fix as e621 did recently - on a gallery page, Raccoony should check to see if the site has marked a submission as restricted/blacklisted, before adding that submission to the open-in-tabs list.

Looking at Newgrounds' code, the `a.item-portalitem-art-small` for the art thumbnail contains a couple of divs that might be useful to key on:

```
<div class="rated-a item-suitability"></div>
  <div class="item-details error text-align-center strong">Restricted 18+</div>
```

Similarly, the movie thumbnail for has a similar `<div class="item-suitability background-color-rated-a">`, but in a slightly different place in the DOM.  The movie thumbnail also has a `span.rated-a` that might be a thing to key on.

#### Observations:

1.  The plugin doesn't seem to run on an artist's "top level" page.  You have to click the "More" button on the "Latest Art" box on their page to get a larger thumbnail gallery, and then the plugin runs.  Example: The plugin doesn't run on (sfw) https://vinnx4.newgrounds.com/ , but does run on (sfw) https://vinnx4.newgrounds.com/art/ .

2.  Newgrounds apparently lets the artist include images in the description, along with the text.  At least one artist seems to use this to include similar images they have drawn before.  It doesn't appear to be a multi-image submission (does Newgrounds even support those?) - there are just some images in the comments.

The main submission image, if static, will have a click-to-embiggen feature, but the comment images won't.  (The main submission image does not have click-to-embiggen if it's WebP.)  The main submission image will have a URL starting with `https://art.ngfiles.com/images/`.  The description images will have a URL starting with `https://art.ngfiles.com/comments/`.

Example, sfw: https://www.newgrounds.com/art/view/vinnx4/diavlo

3.  The infinite-scroll detection kept me from doing something I wanted to do.  I was on this page (sfw) https://flooko.newgrounds.com/art/ , selected "open all in tabs", and Raccoony duly opened eight new tabs with the submissions in them.  I then decided I wanted to auto-download those eight submissions, so I closed the submission tabs, turned on auto-download, and tried "open all in tabs" again.  I got the "all these links have been opened, scroll down to load more" pop-up.

Reloading the gallery page fixed it - the "open all in tabs" button opened eight tabs without complaint.  I feel this might sometimes be confusing to users, though.

4.  The first time I bring up Raccoony's settings menu (click on raccoon head, click on the gear icon), on a page where the raccoon head appears, I get the following complaint in the console:

```
Warning: Failed prop type: You provided a `checked` prop to a form field without an `onChange` handler. This will render a read-only field. If the field should be mutable use `defaultChecked`. Otherwise, set either `onChange` or `readOnly`.
    in input (created by SettingsRow)
    in td (created by SettingsRow)
    in tr (created by SettingsRow)
    in SettingsRow (created by SiteSettingsUi)
    in tbody (created by SiteSettingsUi)
    in table (created by SiteSettingsUi)
    in div (created by SiteSettingsUi)
    in SiteSettingsUi (created by SettingsUi)
    in div (created by SettingsUi)
    in SettingsUi (created by PageOverlay)
    in div (created by PageOverlay)
    in div (created by PageOverlay)
    in PageOverlay (created by Page)
    in div (created by Page)
    in Page page_inject.js:2601:15
```
It doesn't seem to affect anything, and I don't remember if this complaint happened before, but I figured I'd note it.  I'm pretty sure `SettingsUI` and `SiteSettingsUI` are coming from Raccoony.

5.  I Googled and it is apparently *possible* for a submission at Newgrounds to have two artists, and possibly more, listed on the submission.  All of the artists listed have to have a Newgrounds account.  I haven't yet been able to find a submission like that, but I am still looking.

6.  "Auto download opened submissions" won't do anything if you do an open-all-in-tabs on movies.  I don't think there's a way to fix this, but it should probably be documented somewhere.

#### Questions:

1.  Should the links to the description images get included when the plugin grabs the comment text?  This mainly affects what shows up in the metadata file.  Or, should we try to download the description images as well, in case the artist is using them to have multiple images on one submission?

2.  If you spend a while scrolling down on one of the gallery pages (like the "new art", "popular art", or a search-results page with a lot of hits), Raccoony will duly follow the infinite scroll, and keep adding submissions to his "open in tabs" list.  If the user then selects "open in tabs", Raccoony could potentially try to open hundreds and hundreds of tabs.  Should there be some kind of (configurable?) limit, beyond which Raccoony pops up a dialog that says "You're trying to open 500 new tabs, are you sure Y/N"?

The fun thing here is that everybody has a different idea of what "too many tabs" means.  Somebody with a $5,000 PC from yesterday might think 1000 tabs is too many, while somebody with a 10-year-old PC might think 50 is too many.  That's why the limit might need to be configurable.

As a simpler (?) step in this direction, would it be possible for the plugin to update the UI as it accumulates submissions, such that the button says "Open all 3 in tabs", or "Open all 539 in tabs", or similar?

If Raccoony eventually gains the ability to stop a running open-all-in-tabs operation, this probably wouldn't need to exist.

3.  Newgrounds shows a "licensing" infobox for most submissions.  Would it be useful to try to scrape this, and include it in at least the metadata file?  Most of them seem to be Creative Commons licenses, so maybe it would be enough to use their ID codes, like `License: cc-by-sa` or `License: cc-by-nc-sa`.

I think one or two other sites may also support license information.  The biggies, like FA and e621, don't, though.  Also, you could argue that since we save the submission URL in the metadata file, anybody interested in licensing or contacting the artist could just use that URL to do so.

4.  I didn't test this, but I thought of it.  In the following scenario, what happens?

You're sitting on a gallery page with plenty of submissions - say, 40.
You tell Raccoony to "open all in tabs", and he starts doing that.
While the open-in-tabs is running, you scroll down on the gallery page, far enough for the infinite scroll to kick in, so Newgrounds appends another batch of submissions to the page - say, 40 more.
Raccoony (presumably) notices the new batch and adds them to his list.
You tell Raccoony to "open all in tabs" again.

Does he open just the 40 new ones?  Or does he try to open all 80 (old+new)?  Or...?

(This may be what the "have we opened this one already" logic that I ran into is for, but I figured I'd ask.)

#### Conclusion:

Hopefully that's enough to chew on for a while.  :D

**Comment by Simon-Tesla** [comment]
Bugs:
1. Good catch. Should at least be a relatively easy fix.
2. Hm, odd. Would have to investigate. That said, auto-download is probably my least tested feature.
3. Ditto
4. Ah, interesting. The thought entered my mind that I should probably check that, but I forgot. Hopefully not too hard to implement.

Observations:
1. Hm. I suspect the code I wrote to handle favorites should work on such pages, will need to investigate tho.
2. Description is only used for the metadata file and is explicitly text only. Gonna leave description support as-is, especially since I suspect that there's no expectations surrounding authorship of the images in a description. I'm fine with right-click-save being our escape hatch here.
3. Reloading the page is the appropriate fix for that issue; the intention is for the "what have you already opened" tracking to be ephemeral. I don't believe there's a good way to detect that all the images have been loaded, which would be a requirement for showing a different message, and making the message longer is just going to lead to people not reading it even more. I don't think there's a better implementation here that isn't more complex than it's worth.
4. Known issue. The warning is incorrectly detecting an issue as I use onClick handlers to manage the state instead of onChange. I don't care enough to bother with it.
5. Huh. Guess we'll see. I suspect that it will either detect the first artist or detect all of the artists as one thing, either of which I'd consider relatively acceptable.
6. Auto-download won't do anything with movies. Wontfix, but sure, don't mind documenting it somewhere. Maybe a FAQ entry.

Questions
1. Descriptions are text-only by design, and that is the behavior on other sites that have non-text content in them. I'm not convinced that images in the description are necessarily always the author's, and right-click-save can handle them in any case. Wontfix.
2. Hm. We could cap it out at, say, 100 at a time and show a message that the user will need to open tabs again for the next batch. I think that's better than a simple confirmation as it reuses the existing logic to behave more like it would if the user was clicking open tabs as they went instead of loading the entire list. (Open tabs queue management UI is on the to do list, though it'll likely be a while yet. The TaskQueue class was a step in that direction.)
3. Maybe? As far as I'm aware, people just use the metadata files for searching for pics on their local machine, so I don't know that it's all that useful in terms of metadata for that.
4. Not an issue. The list of links for open all in tabs isn't generated until you click the button, and that generated list is passed to Raccoony's background script for processing, and it handles the delay and opening the tabs. There is no continuous background process that adds links as they come in; a new list of links isn't generated until the next time the open all in tabs button is pressed. This is true even when you have a mutation selector; all that does is allow the plugin to detect that it can do something and make the UI visible in response to DOM mutations that happen after the page DOM finishes loading but before scripts have a chance to populate the UI further. (The Newgrounds plugin doesn't have or need a mutation selector though.)

**Comment by Simon-Tesla** [comment]
Also, note to self: given that infinite scroll makes things weird, I'll set the sortable boolean to false for all lists even though technically Newgrounds uses serially-numbered IDs. Otherwise you end up with weird starting points for the tabs that are opened which I think will be more confusing than not.

**Comment by Eupeptic1** [comment]
#### Replies to existing stuff

[Bug 2 - open-all-in-tabs not downloading any MP3s] 
> 2. Hm, odd. Would have to investigate. That said, auto-download is probably my least tested feature.

[Bug 3 - open-all-in-tabs only downloading *some* submissions from one artist]
> 3. Ditto

For fun, I tried both of these again, in two different conditions:

Tab interval 5 seconds (as opposed to default 1), not logged in to Newgrounds.
Tab interval 5 seconds, logged in to Newgrounds.

I thought maybe my browser was "too busy" with a 1-second tab interval, which might have been causing the plugin to miss things.  I also wondered if maybe Newgrounds gives "priority" to logged-in users, when serving art or music from their database.

Result: It didn't have any effect on either bug in either condition.

On bug 2 (doesn't auto-download MP3s), I observed that when a new tab opens, there is a rectangular box where the waveform is *going* to be.  The top half of the box is a slightly different color black or grey than the bottom half of the box, and there is a grey `wf` in the upper left corner.  Less than a second after the tab opens, the waveform fills in, in that rectangular box.

On bug 3 (doesn't auto-download *some* submissions), the submissions that *did* auto-download were exactly the same ones as yesterday.  So at least it's repeatable.

I kind of wonder if a cheap fix would be to wait (say) 0.5 second after the new tab opens before scraping it; it *appears* to me that there is sometimes a slight delay before Newgrounds loads all the stuff for the page, which might be why we miss some things.  This delay might cause hard-to-debug problems in the future, though.  The Cadillac fix is probably a mutation selector.

> [Observation] 5. Huh. Guess we'll see. I suspect that it will either detect the first artist or detect all of the artists as one thing, either of which I'd consider relatively acceptable.

See below, under "New stuff".

> [Question] 3. Maybe? As far as I'm aware, people just use the metadata files for searching for pics on their local machine, so I don't know that it's all that useful in terms of metadata for that.

If any third-party tools (like PostyBirb) can use our metadata files, that tool might be interested in having the license info in the metadata, so it can set that field appropriately when posting to sites that support it.

The use case I have in mind is an artist wanting to duplicate their gallery to a new site.  They download their own stuff from their current site, using Raccoony with metadata files turned on.  Then they fire up PostyBirb or equal, point it at the directory of art and metadata files on their local disk, and tell it "repost all this stuff to the new site" - that way they don't have to type in the titles, descriptions, rating, etc, all over again.  On the other hand, third-party tools may *already* support "duplicate gallery from site 1 to site 2" on their own, without having to use Raccoony to help.

Other points:

I understand and agree with everything else.

#### New stuff

Bug 5: Movie downloads don't seem to work for Flash submissions.

If you click "Download" on the static intro page for a Flash submission, you get the "must click play" alert, as expected.

If you then click "Play" on Newgrounds, Newgrounds loads the Ruffle flash emulator, and then loads the original Flash submission in Ruffle.

If you click "Download" again, you still get the "must click play" alert, which is not expected.  The best expected action is for Raccoony to start downloading the Flash file.  A possible expected action is for Raccoony to give a "no Flash, sorry" alert.

Example, NSFW:  https://www.newgrounds dot com/portal/view/466650

Analysis:

The link to the Flash file seems to be available, even when you're sitting on the static screen for the movie, before you click Play on Newgrounds.  The last few entries in the CSS path to the movie display area are `div.pod.embed div#embed_podcontent.pod-body.breakout div#embed_wrapper div#embed_sizer div.text-align-center iframe#ruffle_embed` .

The `iframe` has a `src=` with a long URL that directly references the Ruffle emulator.  One of the parameters to Ruffle is a JSON-like array; one of the elements of that array is named `swf` and contains the URL for the Flash file.  The Flash file URL resembles `https://uploads.ungrounded.net/123000/123456_myniftymovie.swf`.

Once you click Play on Newgrounds, Ruffle seems to be doing most of its stuff in a `<canvas>` inside the iframe.  There doesn't seem to be a link to the Flash file in the HTML inside the iframe, but the iframe tag itself still has the src that gives you the Flash file link.

Bug (maybe) 6: On submissions with multiple authors, Raccoony only records the first author in the metadata file.

I discovered the "animation collaborations" page, https://www.newgrounds.com/collection/animationcollabs , which has plenty of submissions with multiple authors.  I tried downloading one, with metadata turned on.  Raccoony saw the first artist listed, and the directory in `~/Downloads/raccoony` was named for that artist.  In the metadata file, only the first artist was listed.

Even if the directory is only named for the first artist, it might be good if we could list all the artists in the metadata file.

Example, rated "T" by Newgrounds: https://www.newgrounds dot com/portal/view/818492

There's also an "art collaborations" page at https://www.newgrounds.com/collection/artcollabs .  Most of the submissions there are either games (which we don't handle) or old enough to be Flash, so they run into bug 5.  But it might be useful for more examples of the markup for multiple authors.

Analysis:

The `getCreator()` function in the plugin is looking for a single instance of `.item-user`.  On submissions with multiple authors, there is an instance of `<div class="item-user">` for each author, one instance right after the other, in the "Credits & Info" box.  Maybe a `querySelectorAll()` and a join would do the trick.

**Comment by Simon-Tesla** [comment]
**Autodownload:** If a delay like that, or something like waiting for the tab to become the active tab, makes it work, then I'm for it. I'm generally ok with v1 of autodownload support being rough tho, so if the fix is harder than that I'm not going to hold up a release over it.

**Metadata files:** Certainly I did have some idea in mind for figuring out a PostyBirb integration at some point, tho I'd want to collaborate with PostyBirb's author directly before figuring out exactly what it needs support-wise. I don't recall PostyBirb having a license field in its templates. This is probably not hard to implement though. Like I said, we'll see, not a high priority for me at the moment tho.

**Flash support:** I'm very okay with flash just being unsupported for V1 if it's too difficult to get the file URL; there are other projects that have done the work to archive the flash files on Newgrounds, and it's very difficult for end-users to play the flash files in any case. But it sounds like it might not be too hard to detect and grab the file.

**Multiple authors:** Raccoony's schema doesn't really support the notion of multiple authors, so it'd have to be something like concatenating all of the names together with commas or something, yeah.

**Comment by Simon-Tesla** [comment]
(Also, if you'd like to work on any of this, feel free; I haven't done any additional work on the Newgrounds plugin to address bugs yet.)

**Comment by Eupeptic1** [comment]
> **Flash support:** I'm very okay with flash just being unsupported for V1 if it's too difficult to get the file URL; there are other projects that have done the work to archive the flash files on Newgrounds, and it's very difficult for end-users to play the flash files in any case.

I was thinking about it because for many years, Newgrounds' entire claim to fame was Flash games and animations.  They even distribute their own [standalone Flash player](https://www.newgrounds.com/flash/player) to help deal with the end of browser support.  I will see if I can pull the URL out of the iframe source link.

> **Multiple authors:** Raccoony's schema doesn't really support the notion of multiple authors, so it'd have to be something like concatenating all of the names together with commas or something, yeah.

The e621 plugin sort of accidentally supports multiple authors.  e621 concatenates them for us in the tag we look at, so we see something like `coolartist1 and niftydrawer2` as the author, which turns into `Downloads/raccoony/e621/coolartist1_and_niftydrawer2` on the filesystem.

Do you think `_and_` is a good joiner for Newgrounds, or would something else be preferable?

I won't have time the rest of today (Sunday, 10-02) to work on this, but I can probably work on it early this week (week of 10-3).

**Comment by Simon-Tesla** [comment]
Ah, interesting, forgot about e621, tho it looks like we just found what e621 generates for multiple authors and use that directly instead of doing our own concatenation. I'd personally prefer a comma assuming that most filesystems can take commas in names as it's shorter and easier to localize if I ever get off my butt to add localization support to Raccoony. :P But I don't have a strong opinion on it either.

**Comment by Eupeptic1** [comment]
**Flash support:** I modified the plugin to pull the URL out of the iframe src link. I have tested it with three Flash movies/games and it seems to work OK.

While working on this, I noticed that when downloading a "modern" movie (.mp4 or similar), Raccoony was including Newgrounds' cache-buster/nonce at the end of the URL in the `url` and `previewUrl` fields in the metadata file.  I fixed it to trim the cache-buster off of the URL, for both modern and Flash movies.

**Multiple authors:** I modified the plugin to scrape all authors from the page.  If there is one author, it works as before.  If there is more than one, it joins all their names with a comma, and uses that string in the metadata file and on the filesystem.  I have tested this with three movies.

**Autodownload:** I plan to work on this in the next couple of days.

**Comment by Eupeptic1** [comment]
**tl;dr:** I fixed a bunch of stuff, but still have a few things to go.  My latest is up at https://github.com/Eupeptic1/RaccoonyWebEx/tree/newgrounds-simon , specifically https://github.com/Eupeptic1/RaccoonyWebEx/blob/newgrounds-simon/app/plugins/newgrounds.ts , if you want to review it or work on anything.

**Autodownload:** I worked on this tonight.  I tried various time delays, but nothing really helped.  I took the delays out, and decided to look at the signalling between "main" Raccoony and the plugin.  I discovered that [`checkFileDownload()`](https://github.com/Simon-Tesla/RaccoonyWebEx/blob/master/app/plugins/newgrounds.ts#L64) was *not implemented* in the plugin for video or audio files, which was causing auto-download to not work.  Changing that function to check all three submission types made auto-download of `.mp3` files start working.

The other autodownload bug - where Raccoony repeatably auto-downloaded some, but not all, of the images in one artist's gallery - turned out to be related to the next bug; fixing that bug allowed Raccoony to auto-download that artist's complete gallery.

**Not finding WebP:** This turned out to be a bit of a misnomer.  The markup difference depends on the *pixel size* of the image, not the image type; I just happened to see it first with WebP files.  The pixel size controls whether Newgrounds offers click-to-embiggen or not:

* If any type of image is small enough to fit in the space on the page, Newgrounds just displays it, without click-to-embiggen.
* If a non-WebP image is larger than the space on the page, or if a WebP image is much larger than the space on the page, Newgrounds scales the image down to fit, with click-to-embiggen.
* If a WebP image is a little bit larger than the space on the page, Newgrounds scales it down to fit, without click-to-embiggen.

I changed `getArtMedia()` to find the image correctly in either case.

I also noticed that when downloading art, Raccoony was including Newgrounds' cache-buster/nonce at the end of the URL in the `url` and `previewUrl` fields in the metadata file. I fixed it to trim the cache-buster off of the URL for art.

**Blacklist:** TODO, probably in the next couple of days.

I like what I have enough that I put it up on a new branch in my repo, https://github.com/Eupeptic1/RaccoonyWebEx/tree/newgrounds-simon  .

**Comment by Simon-Tesla** [comment]
I went ahead and created a pull request so that I could put my comments inline. You should be able to continue checking in on your branch and the pull request should update with those changes.

**Comment by Simon-Tesla** [comment]
Also, overall looking good so far, honestly the main thing is that you'll want to use the built in URL and JSON parsing capabilities of the Web API instead of doing brittle stuff with plain text manipulation.

**Comment by Simon-Tesla** [comment]
Also, feel free to open your own pull request if you'd like. That way it tracks you as the owner and such; if I open it, I'm the 'owner' and it'll likely check in under my name.

**Comment by Eupeptic1** [comment]
> Also, feel free to open your own pull request if you'd like.

I can open a PR, but I prefer to wait until I am closer to "done" to do that.  There isn't a real formal process here, but to me, opening a PR means "I think this stuff is pretty likely to work right and be ready for production", and I'm not at that point with this code yet.

I'm not too particular about what name it ends up under.  When you do a release, and all the WebExtension groupies give you flowers and champagne, save one bottle for me.  :)

**Comment by Simon-Tesla** [comment]
Fair enough. I'm used to using pull requests at work for in-progress stuff as I find it *way* more convenient to comment on code with one, though at work they use Azure Dev Ops which has a more formal process for marking a PR as a draft and the approval process is more rigorous than what GitHub does by default.

I'd say if you put something like 'WIP' or 'Draft' in the title, I'm pretty unlikely to mistakenly merge it :P

**Comment by Simon-Tesla** [comment]
Actually, you *can* mark PRs as drafts in GitHub, and I've done so for this one. You can either do so when creating a PR by choosing 'create a draft pull request' from the dropdown on the 'create pull request' button, or convert it after the fact via the "reviewers" gear.
![image](https://user-images.githubusercontent.com/2214046/194216162-5e4904ae-ce5a-4ef9-82c2-5e0982a05927.png)

Would highly recommend creating PRs as soon as you're ready for feedback, in that case.

**Comment by Simon-Tesla** [comment]
Heya @Eupeptic1, working up the motivation to work on Raccoony again now that I've got a number of other household projects completed. Curious if you want to get that branch ready to merge, or if you'd like me to take a pass at it?

**Comment by Eupeptic1** [comment]
I haven't done anything else with it after whatever the latest change on that branch is.  In particular, I haven't tested it against any changes that Newgrounds may have made since then.

You'll probably need to merge it, if you still want the code.
